### ISSUE TITLE
Resolve or remove TODO 3.0 in ssl/t1_lib.c

### ISSUE THREAD

**Issue by t8m** [openssl/openssl#14333]
There are two of them to be resolved or removed:
```
/* TODO(3.0): An arbitrary amount for now. Take another look at this */
# define MAX_GROUPLIST   40
```

```
        /*
         * Check hash is available.
         * TODO(3.0): This test is not perfect. A provider could have support
         * for a signature scheme, but not a particular hash. However the hash
         * could be available from some other loaded provider. In that case it
         * could be that the signature is available, and the hash is available
         * independently - but not as a combination. We ignore this for now.
         */
```

**Comment by gotthardp** [comment]
The second one is quite crucial (at least) for TPM2-based keys. (With restricted keys) the TPM can only sign hashes produced by is own hash implementation. So, when a key is TPM2-based, the hashes implemented by the default provider are not compatible with the TPM signature algorithm. At this moment this causes an "internal error" during TLS negotiation.

**Comment by t8m** [comment]
Hmm, but I think this is actually a slightly different although related issue. Or rather that this adds another level of complexity to the problem. But thank you for that remark, it is quite important to realize that.

**Comment by gotthardp** [comment]
Oh, I thought the remark is talking about the same. Sorry for the misunderstanding. I just found that the hash compatibility is checked elsewhere.
https://github.com/openssl/openssl/blob/e60e974414a7e637ff2f946dc2aa24c381a32cc2/ssl/t1_lib.c#L3028-L3034

**Comment by mattcaswell** [comment]
Fix for the hard coded group list limit in #14886.

The other problem is much harder to solve. I'm proposing to defer this to post 3.0 and have raised a new issue for it (#14885).
